Update TF/TFL -> TOSA legalization tests with SSA values associated

armkevincheng opened the pull request:
This is per https://github.com/tensorflow/tensorflow/pull/48193 requested.

Updating existing tests with SSA values to construct a dependency graph.
This is crucial to check complicated lowering like gather, softmax.

armkevincheng committed:
Update TF/TFL -> TOSA legalization tests with SSA values associated

Signed-off-by: Kevin Cheng <kevin.cheng@arm.com>

armkevincheng commented:
Added @stellaraccident and @rsuderman to the reviewer list

armkevincheng commented:
Previous comment doesn't tag successfully... let me try again.
Add @stellaraccident and @rsuderman to reviewer list.

rsuderman commented on tensorflow/compiler/mlir/tosa/tests/tf-to-tosa-pipeline.mlir:
```diff
@@ -105,13 +105,13 @@ func @test_relu6(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
+// CHECK: %[[VAR3:.*]] = "tosa.mul"(%arg0, %[[VAR2]]) {shift = 0 : i32}
+// CHECK: %[[VAR4:.*]] = "tosa.reshape"(%[[VAR0]]) {new_shape = [1, 1]}
+// CHECK: %[[VAR5:.*]] = "tosa.greater_equal"(%arg0, %[[VAR4]])
+// CHECK: %[[VAR6:.*]] = "tosa.select"(%[[VAR5]], %arg0, %[[VAR3]])
```
Comment regard CHECK vs CHECK-DAG. CHECK-DAG allows different ordering of operations, so it should be preferred in most cases, though the root nodes on the compute graph need to be CHECK statements. In this case, only the "tosa.select" should be a CHECK, the rest should be CHECK-DAGs.

This is needed a many of the checks below.

armkevincheng replied:
Thanks. Just apply the fix to all other tests in commit#2

rsuderman commented on tensorflow/compiler/mlir/tosa/tests/tfl-to-tosa-pipeline.mlir:
```diff
@@ -2,673 +2,662 @@
   %cst = constant dense<0.000000e+00> : tensor<16xf32>
-  %0 = "tfl.conv_2d"(%arg0, %cst_0, %cst)  {dilation_h_factor = 1 : i32, dilation_w_factor = 1 : i32, fused_activation_function = "NONE", padding = "SAME", stride_h = 1 : i32, stride_w = 1 : i32}  : (tensor<1x32x32x8xf32>, tensor<16x1x1x8xf32>, tensor<16xf32>) -> tensor<1x32x32x16xf32>
-  return %0 : tensor<1x32x32x16xf32>
+    %0 = "tfl.conv_2d"(%arg0, %arg1, %cst)  {dilation_h_factor = 1 : i32, dilation_w_factor = 1 : i32, fused_activation_function = "NONE", padding = "SAME", stride_h = 1 : i32, stride_w = 1 : i32}  : (tensor<1x32x32x8xf32>, tensor<16x1x1x8xf32>, tensor<16xf32>) -> tensor<1x32x32x16xf32>
```
This file has inconsistent indentation throughout. Multiple cases of 8+ spaces before a return statement. Please remove the extra indentation (2 spaces per nesting).

armkevincheng replied:
Fixed in commit#2

rsuderman reviewed (changes requested):
Mostly good, just need some fixes for indentation and CHECK vs CHECK-DAG.

armkevincheng committed:
Fixes in response to Rob's feedback on PR#48573

1. Replacing CHECK with CHECK-DAG if it's not root node.
2. Fix indent that I accidentally add to tfl-to-tosa-pipeline.mlir.

rsuderman reviewed (changes requested):
Imported and tfl-to-tosa-pipeline.mlir appears to be failing.

armkevincheng commented:
What I'm running now is:
`bazel test --config=v2 tensorflow/compiler/mlir/tosa/... -j 16`
and this runs fine

Is there any way I can reproduce the failure on my side?

rsuderman commented:
> What I'm running now is:
> `bazel test --config=v2 tensorflow/compiler/mlir/tosa/... -j 16`
> and this runs fine
> 
> Is there any way I can reproduce the failure on my side?

I'm not sure why but bazel on the github repo is not reproducing the issue. I patched in some changes and should be able to land.

For interest sake, it was related to test_matmul not having the correct checks. A few lines of checks were missing.
